lib/phoenix_live_view/tag_engine.ex

```diff
@@ -1190,131 +1190,64 @@ defmodule Phoenix.LiveView.TagEngine do
       tag_meta[:for] || raise_syntax_error!("cannot use :key without :for", tag_meta, state)
 
     if_expr = tag_meta[:if]
-    ast = invoke_subengine(state, :handle_body, [[skip_require: true, root: true]])
+    for_assigns = extract_variables(for_expr)
+    vars_changed = Enum.map(for_assigns, fn {name, _} -> name end)
 
-    try do
-      map_assigns(for_expr, key_expr, ast)
-    rescue
-      e in ArgumentError ->
-        raise_syntax_error!(Exception.message(e), tag_meta, state)
-    else
-      {for_expr, for_assigns, key_expr, assigns_to_pick} ->
-        new_ast =
-          quote do
-            Phoenix.LiveView.TagEngine.component(
-              &Phoenix.Component.live_component/1,
-              %{
-                id: unquote(key_expr),
-                module: Phoenix.LiveView.KeyedComprehension,
-                render: fn var!(assigns) -> unquote(ast) end,
-                assigns:
-                  Map.merge(Map.take(var!(assigns), unquote(assigns_to_pick)), %{
-                    unquote_splicing(for_assigns)
-                  })
-              },
-              {__MODULE__, __ENV__.function, __ENV__.file, unquote(tag_meta.line)}
-            )
-          end
-
-        state = pop_substate_from_stack(state)
-
-        for_ast =
-          state
-          |> push_substate_to_stack()
-          |> update_subengine(:handle_begin, [])
-          |> update_subengine(:handle_expr, ["=", new_ast])
-          |> invoke_subengine(:handle_end, [])
-
-        if if_expr do
-          quote do
-            for unquote(for_expr), unquote(if_expr), do: unquote(for_ast)
-          end
-        else
-          quote do
-            for unquote(for_expr), do: unquote(for_ast)
-          end
-        end
-    end
-  end
+    ast =
+      invoke_subengine(state, :handle_body, [
+        [skip_require: true, root: true, vars_changed: vars_changed]
```

Comment:
@josevalim storing it in vars is good, but we cannot store `:change_track` in `meta` because we actually process the body before we process the for. We could change that, but then we'd need to change the engine to handle the keyed comprehension directly, and I think it's fine to have most of the knowledge of keyed comprehensions in the tag engine instead. So we still need to pass the vars_changed here.